### ISSUE TITLE
Internal hashtable indexing performance problem

### ISSUE THREAD

**Issue by dktapps** [pmmp/PocketMine-MP#2675]
### Issue description
Integer hashed keys are internally used to index chunk and blockcache hashtables in the Level. The problem with these hashes is that PHP uses them as actual underlying hashtable indexes, which causes some major issues given that the hash values are separated by huge distances for small spatial distances.

The below script (simulating block cache indexing) can be used to verify the problem. Notice that string keys (which are internally hashed and treated differently) always take approximately the same time per write or read, while the read/write time goes up exponentially for integer keys.

<details><summary>benchmark.php</summary>

```php
<?php

function hash1($x, $y, $z){
	return (($x & 0xFFFFFFF) << 36) | (($y & 0xff) << 28) | ($z & 0xFFFFFFF);
}

function hash2($x, $y, $z){
	return pack("NCN", $x, $y, $z) . "!"; //trailing character to prevent coercion to integer in case encoded hash is a valid numeric string
}

for($loops = 1; true; ++$loops){
	echo "--- $loops x 256 x $loops ---" . PHP_EOL;
	$a1 = [];
	$a2 = [];

	$start = microtime(true);
	for($x = 0; $x < $loops; ++$x){
		for($y = 0; $y < 256; ++$y){
			for($z = 0; $z < $loops; ++$z){
				$a1[$h = hash1($x, $y, $z)] = $h;
			}
		}
	}
	$time = (microtime(true) - $start);
	echo "int hash write: " . ($time * 1000) . " ms (" . round(($time * 1000000000) / count($a1)) . " ns per write)" . PHP_EOL;

	$start = microtime(true);
	for($x = 0; $x < $loops; ++$x){
		for($y = 0; $y < 256; ++$y){
			for($z = 0; $z < $loops; ++$z){
				$a2[$h2 = hash2($x, $y, $z)] = $h2;
			}
		}
	}
	$time = (microtime(true) - $start);
	echo "string hash write: " . ($time * 1000) . " ms (" . round(($time * 1000000000) / count($a2)) . " ns per write)" . PHP_EOL;

	if(count($a1) !== count($a2)){
		die("array sizes are wrong: " . count($a1) . " vs " . count($a2));
	}

	$start = microtime(true);
	for($x = 0; $x < $loops; ++$x){
		for($y = 0; $y < 256; ++$y){
			for($z = 0; $z < $loops; ++$z){
				$vv1 = $a1[$h = hash1($x, $y, $z)];
				assert($vv1 === $h);
			}
		}
	}
	$time = (microtime(true) - $start);
	echo "int hash read: " . ($time * 1000) . " ms (" . round(($time * 1000000000) / count($a1)) . " ns per read)" . PHP_EOL;

	$start = microtime(true);
	for($x = 0; $x < $loops; ++$x){
		for($y = 0; $y < 256; ++$y){
			for($z = 0; $z < $loops; ++$z){
				$vv2 = $a2[$h = hash2($x, $y, $z)];
				assert($vv2 === $h);
			}
		}
	}
	$time = (microtime(true) - $start);
	echo "string hash read: " . ($time * 1000) . " ms (" . round(($time * 1000000000) / count($a2)) . " ns per read)" . PHP_EOL;
	sleep(1);
}
```
</details>

### Possible solutions
- Make all hashtable indexes use strings instead of integers. This has been demonstrated to alleviate the problem.
- Use [Z-order curve](https://en.wikipedia.org/wiki/Z-order_curve) to create integer hashes. This has been demonstrated to relieve the problem, but comes with extra complexity cost and appears to be slower than simply using strings.

**Comment by dktapps** [comment]
This has additional complications in that any alternative indexing function is guaranteed to be slower than the one we're currently using due to added complexity. The binary keys test has a significant impact on the key computation speed - 100% slower for key generation and 300% for decoding the key back into its original components.

**Comment by dktapps** [comment]
It turns out that this problem is having an enormously negative impact on `LightUpdate` performance, which uses `blockHash` to index its `removalVisited` and `spreadVisited` arrays. I measured an impact of 40-45% performance loss compared to using string keys or Z-order codes.

Some technical detail for this problem: PHP hashtables (like most hashtables) use arrays of linked lists to store their values. A hash is taken of the key (usually someArbitraryHash() % nTableSize), and then the value is added to the corresponding linked list for that key.
If you're lucky, you get the list to yourself, which means that your value can be quickly found. If there's a hash collision, your value gets stuck in a list with a bunch of other values that had the same hash.

PHP uses integer keys directly without passing them through a hash function. `blockHash()` and `chunkHash()` produce hashes that always have the same least-significant-bits for a given Z coordinate, so all values that share a Z coordinate get dumped into the same linked list. With large value sets, this can be extremely problematic, because the list has to be brute-force searched to find the correct value in a collision.

Profiling in Visual Studio shows that in my benchmark above, 97.9% of the CPU time is spent in `_zend_hash_index_find()`, which calls [this inline function](https://github.com/php/php-src/blob/04441585292e4c75d57a2679c62766701b94ce81/Zend/zend_hash.c#L712) which loops over the linked list for a given key. This is called for both read and [write](https://github.com/php/php-src/blob/04441585292e4c75d57a2679c62766701b94ce81/Zend/zend_hash.c#L1043), explaining why the performance hit applies on both reads and writes to arrays.

Any hash function which improves the distribution of the least significant bits would solve this problem; however, Morton codes are ideal for our use case because the problem directly comes from multi-dimensional indexing and Morton codes provide memory locality for stuff that's local on dimensions.

**Comment by dktapps** [comment]
This has been partially addressed by 0f9d5f70113fded96aa823d2881f1edc9914f987, although blockHash() remains to be addressed (the main issue being loss of resolution on the X/Z axes).

**Comment by dktapps** [comment]
This has now been fully resolved as of 286ac2a97597a1d52424ec77a9e17e4155f190d9.
